src/backend/distributed/planner/intermediate_result_pruning.c

```diff
@@ -163,10 +164,17 @@ AppendAllAccessedWorkerNodes(List *workerNodeList, DistributedPlan *distributedP
 		foreach(placementCell, task->taskPlacementList)
 		{
 			ShardPlacement *placement = lfirst(placementCell);
+
+			if (placement->nodeId == LOCAL_NODE_ID)
+			{
+				entry->writeLocalFile = true;
+				continue;
+			}
+
 			workerNodeList = list_append_unique_int(workerNodeList, placement->nodeId);
 
 			/* early return if all the workers are accessed */
-			if (list_length(workerNodeList) == workerNodeCount)
+			if (list_length(workerNodeList) == workerNodeCount && entry->writeLocalFile)
```

Comment:
I'm on it. I'll share my findings

Reply:
I created a cluster with 10 nodes, and 256 shards for distributed tables. The below tests are on almost empty distributed tables on my laptop.

For many of the queries in the regression tests, the time to execute `RecordSubplanExecutionsOnNodes()` on `ExecuteSubPlans()` is around 4-5 msecs. Note that this includes all the calls to `AppendAllAccessedWorkerNodes ()` for the execution.

This is obviously a lot if you look the number in isolated. But look into some other information:


- Even if the query hits only locally executable CTEs, the total for `ExecuteSubPlans`  is around 50-100 msecs, and the total execution is more depending on the queries inside CTEs

- With 32 shards and and 10 workers, it is 0 msecs and I cannot measure below msecs precision :) 

- Before the this patch, `RecordSubplanExecutionsOnNodes()` was 3-4 msecs with the same configuration. So, it is fair to say we should consider improving it in general, like using a hashmap or bitmap.

- For the following query, where previously the optimization would kick in, but not in the patch, the call for `RecordSubplanExecutionsOnNodes()` takes almost the same, 27msecs:

```SQL
WITH cte_1 AS (SELECT * FROM table_1),
cte_2 AS (SELECT * FROM table_1),
cte_3 AS (SELECT * FROM table_1),
cte_4 AS (SELECT * FROM table_1),
cte_5 AS (SELECT * FROM table_1),
cte_6 AS (SELECT * FROM table_1),
cte_7 AS (SELECT * FROM table_1),
cte_8 AS (SELECT * FROM table_1),
cte_9 AS (SELECT * FROM table_1),
cte_10 AS (SELECT * FROM table_1)
SELECT count(*) FROM cte_1 JOIN cte_2 USING (key) JOIN cte_3 USING(key) JOIN cte_4 USING (key) JOIN cte_5 USING (key) JOIN cte_6 USING (key) JOIN cte_7 USING (key) JOIN cte_8 USING (key) JOIN cte_9 USING(key) JOIN cte_10 USING (key) JOIN table_2 USING(key);
```

All and all, my suggestion would be to work on getting rid of `append_list_int_unique` and re-run the benchmarks separately. If you're OK with that, I can open an issue to track

Reply:
Lets open an issue to track this.

It seems from these measurements it is correct to conclude the current patch is not significantly changing the time spent here (correct me if that is a wrong conclusion).

Changing this would involve a lot of testing and precise analysis which would take too long for this patch. Having it in a separate issue makes sense to me.

Reply:
> correct me if that is a wrong conclusion

That's my conclusion as well.

> Changing this would involve a lot of testing and precise analysis which would take too long for this patch. Having it in a separate issue makes sense to me.

See https://github.com/citusdata/citus/issues/3420, feel free to comment on as well